From 137ed720372c634648f0a553db51082a59b35ca7 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Sun, 20 Jun 2021 23:36:09 +0300
Subject: [PATCH] [SW-46313] gaudi: security: add r/o open for sram

Add support for enabling the read-only access to the whole SRAM region.
To make this happen, preboot_security_sram_pci_set_ro() function walks
over all the SRAM protecting dynamic RRs and opens the read-only access.

Change-Id: I5fa6a9ac4ddd5b85efb345bc63a83c17f0b1ce4b
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/gaudi/security.c b/gaudi/security.c
index e02573c..d215353 100644
--- a/gaudi/security.c
+++ b/gaudi/security.c
@@ -1981,3 +1981,28 @@
 	preboot_secure_pcie_rr_rd(cfg->rr, cfg->min,
 				  en ? HL_SPI_BASE : cfg->max);
 }
+
+/*
+ * preboot_security_sram_pci_set_ro() function assumes the TPC3 has already been
+ * moved to be secured through the secure level RRs and thus only opens the PCIe
+ * RRs for reading.
+ */
+void preboot_security_sram_pci_set_ro(bool en)
+{
+	int i;
+	u64 min, max;
+	struct preboot_sec_dyn_pcie_rr_cfg *cfg;
+
+
+	for (i = SRAM_BGN_TO_VERS; i <= IMG_END_TO_SRAM_END; i++) {
+		cfg = &preboot_sec_dyn_cfg[i];
+		min = en ? 0 : cfg->min;
+		max = en ? 0 : cfg->max;
+
+		if (cfg->rr >= 0 && cfg->min && cfg->max) {
+			preboot_secure_pcie_rr_wr(cfg->rr, cfg->min, cfg->max);
+			preboot_secure_pcie_rr_rd(cfg->rr, min, max);
+		}
+	}
+
+}
diff --git a/include/security.h b/include/security.h
index 61e2cfb..bbf0740 100644
--- a/include/security.h
+++ b/include/security.h
@@ -22,5 +22,6 @@
 void preboot_security_sp_sram_write_key_hash(u8 *key, size_t len);
 void preboot_security_hbm_pci_open_ro(void);
 void preboot_security_sf_pci_set_ro(bool en);
+void preboot_security_sram_pci_set_ro(bool en);
 
 #endif /* _HL_SECURITY_H_ */
